builder/builder.go

```diff
@@ -0,0 +1,334 @@
+					stepWorkingDir = step.WorkDir
+				}
+			}
+
+			deps, err := b.scrapeDependencies(ctx, volName, step.WorkDir, step.ID, dockerfile, context, tags, buildArgs)
+			if err != nil {
+				errorChan <- errors.Wrap(err, "failed to scan dependencies")
+				return
+			}
+			step.ImageDependencies = deps
+
+			// Modify the Run command if it's a tar or a git URL.
+			if !util.IsLocalContext(context) {
+				// Allow overriding the context from the git URL.
+				if util.IsGitURL(context) {
+					step.Run = replacePositionalContext(step.Run, getContextFromGitURL(context))
+				} else {
+					step.Run = replacePositionalContext(step.Run, ".")
+				}
+			}
+
+			args = b.getDockerRunArgs(volName, step.ID, stepWorkingDir)
+			args = append(args, "docker")
+			args = append(args, strings.Fields(step.Run)...)
+
+		} else {
+			args = b.getDockerRunArgs(b.workspaceDir, step.ID, step.WorkDir)
+			for _, env := range step.Envs {
+				args = append(args, "--env", env)
+			}
+			if step.EntryPoint != "" {
+				args = append(args, "--entrypoint", step.EntryPoint)
+			}
+			args = append(args, strings.Fields(step.Run)...)
+		}
+
+		if b.debug {
+			fmt.Printf("Args: %v\n", args)
+		}
+
+		// TODO: secret envs
+
+		// NB: ctx refers to the global ctx here,
+		// so when running individual processes use the individual
+		// step's ctx instead.
+		timeout := time.Duration(step.Timeout) * time.Second
+		currCtx, cancel := context.WithTimeout(ctx, timeout)
+		defer cancel()
+		if err := b.cmder.Run(currCtx, args, nil, os.Stdout, os.Stderr, ""); err != nil {
+			step.StepStatus = graph.Failed
+			errorChan <- err
+		} else {
+			step.StepStatus = graph.Successful
+
+			// Try to process all child nodes
+			for _, c := range child.Children() {
+				go b.processVertex(ctx, dag, child, c, errorChan)
+			}
+		}
+
+		step.CompletedChan <- true
+	} else if b.debug {
+		fmt.Printf("Skipped processing %v, degree: %v\n", child.Name, degree)
+	}
+}
+
+// PopulateDigests populates digests on dependencies
+func (b *Builder) PopulateDigests(ctx context.Context, dependencies []*models.ImageDependencies) error {
+	for _, entry := range dependencies {
+		if err := b.queryDigest(ctx, entry.Image); err != nil {
+			return err
+		}
+		if err := b.queryDigest(ctx, entry.Runtime); err != nil {
+			return err
+		}
+		for _, buildtime := range entry.Buildtime {
+			if err := b.queryDigest(ctx, buildtime); err != nil {
+				return err
+			}
+		}
+	}
+	return nil
+}
+
+func (b *Builder) queryDigest(ctx context.Context, reference *models.ImageReference) error {
+	if reference != nil {
+		// refString will always have the tag specified at this point.
+		// For "scratch", we have to compare it against "scratch:latest" even though
+		// scratch:latest isn't valid in a FROM clause.
+		if reference.Reference == NoBaseImageSpecifierLatest {
+			return nil
+		}
+
+		args := []string{
+			"docker",
+			"inspect",
+			"--format",
+			"\"{{json .RepoDigests}}\"",
+			reference.Reference,
+		}
```

Comment:
will it be eventually a container based step?

Reply:
Yeah it should be using the Docker CLI's container. I'll fix that.

Reply:
fixed